app/models/term.rb

```diff
@@ -0,0 +1,150 @@
+# frozen_string_literal: true
+
+class Term < ApplicationRecord
+  TEMPORARY_URI_BASE = 'temp:'
+  LOCAL     = 'local'
+  TEMPORARY = 'temporary'
+  EXTERNAL  = 'external'
+
+  TERM_TYPES = [LOCAL, EXTERNAL, TEMPORARY].freeze
+
+  belongs_to :vocabulary
+
+  before_validation :set_uid, :set_uri, :set_uri_hash, on: :create
+  before_save :cast_custom_fields
+  after_commit :update_solr # Is triggered after successful save/update/destroy.
+
+  validates :vocabulary, :pref_label, :uri, :uri_hash, :uid, :term_type, presence: true
+  validates :term_type, inclusion: { in: TERM_TYPES, message: 'is not valid: %{value}' }, allow_nil: true
+  validates :uri,  format: { with: /\A#{URI.regexp}\z/ },
+                   if: proc { |t| t.uri? && (t.term_type == LOCAL || t.term_type == EXTERNAL) }
+  validates :uri_hash, uniqueness: { scope: :vocabulary, message: 'unique check failed. This uri already exists in this vocabulary.' }
+  validates :uid, format: { with: /\A\h{8}-\h{4}-4\h{3}-[89ab]\h{3}-\h{12}\z/ }, allow_nil: true
+  validates :alt_labels, absence: { message: 'is not allowed for temporary terms' }, if: proc { |t| t.term_type == TEMPORARY }
+  validate  :uid_uri_and_term_type_unchanged, :pref_label_unchanged_for_temp_term, :validate_custom_fields
+
+  store :custom_fields, coder: JSON
+
+  serialize :alt_labels, Array
+
+  def to_solr
+    {
+      'uid'           => uid,
+      'uri'           => uri,
+      'pref_label'    => pref_label,
+      'alt_labels'    => alt_labels,
+      'term_type'     => term_type,
+      'vocabulary'    => vocabulary.string_key,
+      'authority'     => authority,
+      'custom_fields' => custom_fields.to_json
+    }.tap do |doc|
+      vocabulary.custom_fields.each do |k, v|
+        doc["#{k}#{Solr::Utils.suffix(v[:data_type])}"] = custom_fields[k]
+      end
+    end
+  end
+
+  def set_custom_field(field, value)
+    custom_fields[field] = value
+  end
+
+  def self.local_uri_prefix
+    host = HYACINTH[:local_uri_prefix]
+
+    raise 'Missing local_uri_prefix in config/hyacinth.yml' unless host
+    host.ends_with?('/') ? host : "#{host}/"
+  end
+
+  private
+
+    def cast_custom_fields
```

Comment:
In the future, I wonder if it might be worth using something like dry_schema to validate and cast custom field types:
https://dry-rb.org/gems/dry-schema/1.4/

Reply:
Yea, I would love to not have all this custom code to do the validation and casting. I don't know if I would want to delay this PR with that change. Maybe it can be a Hyacinth 3.1 addition.

Reply:
Or roll it into typedefs

Reply:
I don't feel like I have a very deep understand of TypeDefs. Could they be used outside of the digital object model?

Reply:
I'll write up a ticket about this, and we can merge this PR

Reply:
TypeDefs might work, but dry-schema could offer a simpler implementation since this isn't deeply nested data.  I'm open to considering both though.

Reply:
If we did want to open up TypeDef use to non-DigitalObjects, we'd want to move it out of the DigitalObject namespace.